// @web/test-runner-commands/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.snapshotPlugin = exports.filePlugin = exports.a11ySnapshotPlugin = exports.sendKeysPlugin = exports.setUserAgentPlugin = exports.emulateMediaPlugin = exports.setViewportPlugin = void 0;
var setViewportPlugin_1 = require("./setViewportPlugin");
Object.defineProperty(exports, "setViewportPlugin", { enumerable: true, get: function () { return setViewportPlugin_1.setViewportPlugin; } });
var emulateMediaPlugin_1 = require("./emulateMediaPlugin");
Object.defineProperty(exports, "emulateMediaPlugin", { enumerable: true, get: function () { return emulateMediaPlugin_1.emulateMediaPlugin; } });
var setUserAgentPlugin_1 = require("./setUserAgentPlugin");
Object.defineProperty(exports, "setUserAgentPlugin", { enumerable: true, get: function () { return setUserAgentPlugin_1.setUserAgentPlugin; } });
var sendKeysPlugin_1 = require("./sendKeysPlugin");
Object.defineProperty(exports, "sendKeysPlugin", { enumerable: true, get: function () { return sendKeysPlugin_1.sendKeysPlugin; } });
var a11ySnapshotPlugin_1 = require("./a11ySnapshotPlugin");
Object.defineProperty(exports, "a11ySnapshotPlugin", { enumerable: true, get: function () { return a11ySnapshotPlugin_1.a11ySnapshotPlugin; } });
var filePlugin_1 = require("./filePlugin");
Object.defineProperty(exports, "filePlugin", { enumerable: true, get: function () { return filePlugin_1.filePlugin; } });
var snapshotPlugin_1 = require("./snapshotPlugin");
Object.defineProperty(exports, "snapshotPlugin", { enumerable: true, get: function () { return snapshotPlugin_1.snapshotPlugin; } });
